cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -1693,7 +1692,9 @@ struct ArithmeticFunction : ScalarFunction {
     if (values->size() == 2) {
       ReplaceNullWithOtherType(values);
 
-      if (auto type = CommonNumeric(*values)) {
+      if (auto type = CommonTemporal(values->data(), values->size())) {
```

Comment:
Now that I think about it, this is kind of weird to use CommonTemporal since there isn't actually a "common" type between date32 and duration (for instance) and what we actually care about here is the common unit - maybe it's worth splitting that off separately? Since future/other uses of CommonTemporal might be 'surprised' to find that `CommonTemporal(date, duration) => date` instead of nullptr.

Reply:
How about renaming `CommonTemporal` to `CommonResolution` and returning finest time unit resolution?

Reply:
CommonTemporal is used elsewhere, but having CommonResolution alongside it makes sense.

Reply:
Added `CommonTemporalResolution` and refactored `ReplaceTemporalTypes`. This PR should hopefully cover all temporal casts needed for [ARROW-11901](https://issues.apache.org/jira/browse/ARROW-11901).